docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -294,254 +296,284 @@ TODO
+
+A transaction may be removed in two scenarios: when updating a list of committed transactions, and
+during rechecking if the transaction is reassessed as invalid. In either case, the first step is to
+identify the lane the transaction belongs to by accessing the `txLanes` map. Then, we remove the
+entry from the CList corresponding to its lane and update the auxiliary variables accordingly.
+
+Since the broadcast goroutines are constantly reading the list of transactions to disseminate them,
+it's important to prioritize the removal of transactions from high-priority lanes first.
+
+When updating the mempool, there is potential for a slight optimization by removing transactions
+from different lanes in parallel. To achieve this, we would first need to preprocess the list of
+transactions to determine the lane of each transaction. However, this optimization has minimal
+impact if the committed block contains few transactions. Therefore, we decided to exclude it from
+the MVP.
+
+### Reaping transactions for creating blocks
+
+Currently, the function `ReapMaxBytesMaxGas(maxBytes, maxGas)` collects transactions in FIFO order
+from the CList until either reaching `maxBytes` or `maxGas` (both of these values are consensus
+parameters).
+
+With multiple CLists, we need to collect transactions from higher-priority lanes first, also in FIFO
+order, continuing with successive lanes in the `sortedLanes` array, that is, in decreasing priority
+order, and breaking the iteration when reaching `maxBytes` or `maxGas`.
+
+The mempool is locked during `ReapMaxBytesMaxGas`, so no transaction will be added or removed from
+the mempool during reaping.
+
+### Transaction dissemination
+
+For broadcasting transactions from multiple lanes, we have considered two possible approaches:
+1. Reserve $N$ p2p channels for use by the mempool. P2P channels have priorities that we can reuse
+   as lane priorities. There are a maximum of 256 P2P channels, thus limiting the number of lanes.
+2. Continue using the current P2P channel for disseminating transactions and implement logic within
+   the mempool to select the order of transactions to put in the channel. This option theoretically
+   allows for an unlimited number of lanes, constrained only by the nodes’ capacity to store the
+   lane data structures.
+
+We choose the second option for its flexibility, allowing us to start with a simple scheduling
+algorithm that can be refined over time. The desired algorithm must satisfy the properties
+"Priorities between classes" and "FIFO ordering per class". This requires supporting selection by
+_weight_, ensuring each lane gets a fraction of the P2P channel capacity proportional to its
+priority. Moreover, it should be _fair_, to prevent starvation of low-priority lanes. Given the
+extensive research on scheduling algorithms in operating systems and networking, we propose for the
+MVP implementing a variant of the [Weighted Round Robin (WRR)](wrr) algorithm, which meets these
```

Comment:
In QoS for IP networks there is [WRED](https://en.wikipedia.org/wiki/Weighted_random_early_detection). I need check if that would be relevant here.

Reply:
The link is broken.

Reply:
Again, is this what the p2p channel does?

Reply:
> The link is broken.

Which link do you refer to? The one on the text, or the one in my comment (both seem to be working for me)

> Again, is this what the p2p channel does?

Yeah, I had the same question when I read this part of Thu. Is priority in p2p channels implementing `wrr` or something else?

Reply:
The link on WRR was broken and then I fixed it. I added a section called Prioritization Logic with a description of the algorithm in P2P channels.